src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs

```diff
@@ -789,6 +789,34 @@ private void ComputePrCurves()
             }
         }
 
+        /// <summary>
+        /// Generates precision recall curve data points.
+        /// </summary>
+        /// <param name="data">The scored data.</param>
+        /// <param name="label">The name of the label column in <paramref name="data"/>.</param>
+        /// <param name="score">The name of the score column in <paramref name="data"/>.</param>
+        /// <param name="probability">The name of the probability column in <paramref name="data"/>, the calibrated version of <paramref name="score"/>.</param>
+        /// <param name="predictedLabel">The name of the predicted label column in <paramref name="data"/>.</param>
+        /// <returns>The precision recall curve data points.</returns>
+        public IDataView GetPrecisionRecallCurve(IDataView data, string label, string score, string probability, string predictedLabel)
```

Comment:
Hi @ganik, thanks for working on this! I view the API in its current form as having a few potential improvements...

The most obvious thing is this return value... an `IDataView`. Probably not a very good transmission format for metrics results.

It may be informative to note how we surface other metrics results in our public API. You'll note that we *do* have `IEvaluator` (which is an internal interface) return metrics via `IDataView` (because, it needs to communicate them in *some* common format, and also because that's one of the types that entry-points uses), but note what we do with them for the sake of our public API.

https://github.com/dotnet/machinelearning/blob/5b22420d28c0cacc9b265d043555b6d11a017b91/src/Microsoft.ML.Data/Evaluators/Metrics/BinaryClassificationMetrics.cs#L87-L98

We actually take the data out, and store them in an actual strongly typed structure, with intellisense, and whatnot.

We *could* have done what you did here -- just sort of take the `IDataView` even though it was one whose structure we knew *exactly* what was in it, called it a day, and said, "here, good luck figuring out what to do with this," but we didn't do that because that would be a needlessly terrible experience. Here we caneasy to use, discoverable API for something as important as metrics, and there was an obvious way to do it, just as there are with this PR. So, I was expecting, and I kind of think people might appreciate something more closely resembling that, with something with actual types, since unlike in some scenarios (general dataflow) the types and column names are well defined.

Note, here is where this data view is composed, for your own reference, so you can see what goes in it.

https://github.com/dotnet/machinelearning/blob/5b22420d28c0cacc9b265d043555b6d11a017b91/src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs#L373-L392

So, a few columns that are always there. We'll need to figure out what's up with these "strat" columns, that is, "stratification" columns (we don't want to expose this as such re: @rogancarr), so perhaps we skip that one, or figure out some other structure that's appropriate. (We'd probably group them, somehow, I don't quite know offhand what that's for. CV?) Anyway, you'll want to figure out what's up with that.

You could imagine some *alternate* structure, whereby this "stuff" is put into a strongly typed structure. In the short term you could copy the results of the `IDataView` out into this structure -- that's a little inefficient but will suffice. Could be as simple as an array of a struct containing the relevant PR points. 

The second problem with the API is a bit less fundamnetal, and probably easier to fix. You will also note that that you are discarding all the other results of evaluation, for no particular reason. What I had suggested in the original issue, and I still think is a good idea, is `EvaluateWithPRCurve` that still produces the metric results (since they're already ***right there*** and being calculated alongside the PR curves!). So, maybe return the metrics, but also have the PR curve structure, whatever it is (not just an `IDataView`!!), as an `out` parameter. Or reverse the two, I don't know.

Reply:
Awesome, thank you for great suggestions as usually @TomFinley

Reply:
I like this idea: `EvaluateWithPRCurve`

One thing that Gani and I talked about was also being able to return the data for an ROC curve too. The original issue (and related v1 scenario) was specifically about a PR curve, but I think it should be on the table to return an ROC curve too. These are the two major curves that people may want to plot.

Also, we may want to think about being able to extend this design to any sort of curve someone may want to plot, as different fields of applications have slightly different target metrics.

---
In reply to: [267839855](https://github.com/dotnet/machinelearning/pull/3039#discussion_r267839855) [](ancestors = 267839855)

Reply:
this should be done. ROC Curve I suggest to do post V1.

---
In reply to: [267921726](https://github.com/dotnet/machinelearning/pull/3039#discussion_r267921726) [](ancestors = 267921726,267839855)

Reply:
How much extra work is an ROC Curve on top of getting the PR Curve? It's weird to have one but not the other from the API.

---
In reply to: [268027498](https://github.com/dotnet/machinelearning/pull/3039#discussion_r268027498) [](ancestors = 268027498,267921726,267839855)

Reply:
there is no code internally that calculates ROC. PR data has been already calculated internally and in this PR i jsut expose it. For ROC i ll need to implement code that calculates it, I expect it to be few days change (including reviews) and should be done in separate PR

---
In reply to: [268340685](https://github.com/dotnet/machinelearning/pull/3039#discussion_r268340685) [](ancestors = 268340685,268027498,267921726,267839855)